// @volar/server/out/registers/registerlanguageFeatures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const vue = require("vscode-vue-languageservice");
const vscode = require("vscode-languageserver");
function register(features, legend, server, tsVersion) {
    var _a, _b, _c, _d;
    if (features.references) {
        server.referencesProvider = true;
    }
    if (features.definition) {
        server.definitionProvider = true;
    }
    if (features.typeDefinition) {
        server.typeDefinitionProvider = true;
    }
    if (features.callHierarchy) {
        server.callHierarchyProvider = true;
    }
    if (features.hover) {
        server.hoverProvider = true;
    }
    if (features.rename) {
        server.renameProvider = {
            prepareProvider: true,
        };
    }
    if (features.renameFileRefactoring) {
        server.workspace = {
            fileOperations: {
                willRename: {
                    filters: [
                        { pattern: { glob: '**/*.vue' } },
                        { pattern: { glob: '**/*.js' } },
                        { pattern: { glob: '**/*.ts' } },
                        { pattern: { glob: '**/*.jsx' } },
                        { pattern: { glob: '**/*.tsx' } },
                        { pattern: { glob: '**/*.json' } },
                    ]
                }
            }
        };
    }
    if (features.signatureHelp) {
        server.signatureHelpProvider = {
            triggerCharacters: ['(', ',', '<'],
            retriggerCharacters: [')'],
        };
    }
    if (features.completion) {
        const triggerCharacters = vue.getTriggerCharacters(tsVersion);
        server.completionProvider = {
            triggerCharacters: Object.values(triggerCharacters).flat(),
            resolveProvider: true,
        };
        server.executeCommandProvider = {
            commands: [
                ...((_b = (_a = server.executeCommandProvider) === null || _a === void 0 ? void 0 : _a.commands) !== null && _b !== void 0 ? _b : []),
                vue.Commands.CONVERT_TO_KEBAB_CASE,
                vue.Commands.CONVERT_TO_PASCAL_CASE,
            ]
        };
    }
    if (features.documentHighlight) {
        server.documentHighlightProvider = true;
    }
    if (features.documentLink) {
        server.documentLinkProvider = {
            resolveProvider: false, // TODO
        };
    }
    if (features.workspaceSymbol) {
        server.workspaceSymbolProvider = true;
    }
    if (features.codeLens) {
        server.codeLensProvider = {
            resolveProvider: true,
        };
        server.executeCommandProvider = {
            commands: [
                ...((_d = (_c = server.executeCommandProvider) === null || _c === void 0 ? void 0 : _c.commands) !== null && _d !== void 0 ? _d : []),
                vue.Commands.HTML_TO_PUG,
                vue.Commands.PUG_TO_HTML,
                vue.Commands.USE_SETUP_SUGAR,
                vue.Commands.UNUSE_SETUP_SUGAR,
                vue.Commands.USE_REF_SUGAR,
                vue.Commands.UNUSE_REF_SUGAR,
                vue.Commands.SHOW_REFERENCES,
            ]
        };
    }
    if (features.semanticTokens) {
        server.semanticTokensProvider = {
            range: true,
            full: false,
            legend,
        };
    }
    if (features.codeAction) {
        server.codeActionProvider = {
            codeActionKinds: [
                vscode.CodeActionKind.Empty,
                vscode.CodeActionKind.QuickFix,
                vscode.CodeActionKind.Refactor,
                vscode.CodeActionKind.RefactorExtract,
                vscode.CodeActionKind.RefactorInline,
                vscode.CodeActionKind.RefactorRewrite,
                vscode.CodeActionKind.Source,
                vscode.CodeActionKind.SourceFixAll,
                vscode.CodeActionKind.SourceOrganizeImports,
            ],
            resolveProvider: true,
        };
    }
}
exports.register = register;
