// @volar/client/out/nodeClientMain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const lsp = require("vscode-languageclient/node");
const common_1 = require("./common");
function activate(context) {
    return (0, common_1.activate)(context, (id, name, documentSelector, initOptions, port) => {
        const serverModule = vscode.Uri.joinPath(context.extensionUri, 'node_modules', '@volar', 'server', 'node');
        const debugOptions = { execArgv: ['--nolazy', '--inspect=' + port] };
        const serverOptions = {
            run: { module: serverModule.fsPath, transport: lsp.TransportKind.ipc },
            debug: {
                module: serverModule.fsPath,
                transport: lsp.TransportKind.ipc,
                options: debugOptions
            },
        };
        const clientOptions = {
            documentSelector,
            initializationOptions: initOptions,
            synchronize: {
                fileEvents: vscode.workspace.createFileSystemWatcher('{**/*.vue,**/*.js,**/*.jsx,**/*.ts,**/*.tsx,**/*.json}')
            }
        };
        const client = new lsp.LanguageClient(id, name, serverOptions, clientOptions);
        context.subscriptions.push(client.start());
        return client;
    });
}
exports.activate = activate;
function deactivate() {
    return (0, common_1.deactivate)();
}
exports.deactivate = deactivate;
