// @volar/source-map/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceMap = exports.SourceMapBase = exports.Mode = void 0;
var Mode;
(function (Mode) {
    /**
     * @case1
     * 123456 -> abcdef
     * ^    ^    ^    ^
     * @case2
     * 123456 -> abcdef
     *  ^  ^      ^  ^
     * @case3
     * 123456 -> abcdef
     *   ^^        ^^
     */
    Mode[Mode["Offset"] = 0] = "Offset";
    /**
     * @case1
     * 123456 -> abcdef
     * ^    ^    ^    ^
     * @case2
     * 123456 -> abcdef
     *  ^  ^     NOT_MATCH
     * @case3
     * 123456 -> abcdef
     *   ^^      NOT_MATCH
     */
    Mode[Mode["Totally"] = 1] = "Totally";
    /**
     * @case1
     * 123456 -> abcdef
     * ^    ^    ^    ^
     * @case2
     * 123456 -> abcdef
     *  ^  ^     ^    ^
     * @case3
     * 123456 -> abcdef
     *   ^^      ^    ^
     */
    Mode[Mode["Expand"] = 2] = "Expand";
    /**
     * @case1
     * 123456 -> abcdef
     * ^     ^   ^    ^
     * @case2
     * 123456 -> abcdef
     *      ^^   ^    ^
     * @case3
     * 123456 -> abcdef
     *       ^^  NOT_MATCH
     */
    Mode[Mode["Overlap"] = 3] = "Overlap";
})(Mode = exports.Mode || (exports.Mode = {}));
class SourceMapBase {
    constructor(_mappings) {
        this.cache = new Map();
        this.mappings = _mappings !== null && _mappings !== void 0 ? _mappings : [];
    }
    getSourceRange(start, end, filter) {
        for (const maped of this.getRanges(start, end !== null && end !== void 0 ? end : start, false, filter)) {
            return maped;
        }
    }
    getMappedRange(start, end, filter) {
        for (const maped of this.getRanges(start, end !== null && end !== void 0 ? end : start, true, filter)) {
            return maped;
        }
    }
    getSourceRanges(start, end, filter) {
        return this.getRanges(start, end !== null && end !== void 0 ? end : start, false, filter);
    }
    getMappedRanges(start, end, filter) {
        return this.getRanges(start, end !== null && end !== void 0 ? end : start, true, filter);
    }
    *getRanges(startOffset, endOffset, sourceToTarget, filter) {
        const key = startOffset + ':' + endOffset + ':' + sourceToTarget;
        let result = this.cache.get(key);
        if (!result) {
            result = {
                index: 0,
                mapeds: [],
            };
            this.cache.set(key, result);
        }
        for (const maped of result.mapeds) {
            if (!filter || filter(maped.data)) {
                yield getMaped(maped);
            }
        }
        while (result.index < this.mappings.length) {
            const mapping = this.mappings[result.index++];
            const maped = this.getRange(startOffset, endOffset, sourceToTarget, mapping.mode, mapping.sourceRange, mapping.mappedRange, mapping.data);
            if (maped) {
                result.mapeds.push(maped);
                if (!filter || filter(maped.data))
                    yield getMaped(maped);
            }
            else if (mapping.additional) {
                for (const other of mapping.additional) {
                    const maped = this.getRange(startOffset, endOffset, sourceToTarget, other.mode, other.sourceRange, other.mappedRange, mapping.data);
                    if (maped) {
                        result.mapeds.push(maped);
                        if (!filter || filter(maped.data))
                            yield getMaped(maped);
                        break; // only return first match additional range
                    }
                }
            }
        }
        function getMaped(maped) {
            return maped;
        }
    }
    getRange(start, end, sourceToTarget, mode, sourceRange, targetRange, data) {
        const mapedToRange = sourceToTarget ? targetRange : sourceRange;
        const mapedFromRange = sourceToTarget ? sourceRange : targetRange;
        if (mode === Mode.Totally) {
            if (start === mapedFromRange.start && end === mapedFromRange.end) {
                const _start = mapedToRange.start;
                const _end = mapedToRange.end;
                return {
                    data: data,
                    start: Math.min(_start, _end),
                    end: Math.max(_start, _end),
                };
            }
        }
        else if (mode === Mode.Offset) {
            if (start >= mapedFromRange.start && end <= mapedFromRange.end) {
                const _start = mapedToRange.start + start - mapedFromRange.start;
                const _end = mapedToRange.end + end - mapedFromRange.end;
                return {
                    data: data,
                    start: Math.min(_start, _end),
                    end: Math.max(_start, _end),
                };
            }
        }
        else if (mode === Mode.Expand) {
            if (start >= mapedFromRange.start && end <= mapedFromRange.end) {
                const _start = mapedToRange.start;
                const _end = mapedToRange.end;
                return {
                    data: data,
                    start: Math.min(_start, _end),
                    end: Math.max(_start, _end),
                };
            }
        }
        else if (mode === Mode.Overlap) {
            if ((start >= mapedFromRange.start && start <= mapedFromRange.end)
                || (end >= mapedFromRange.start && end <= mapedFromRange.end)
                || (mapedFromRange.start >= start && mapedFromRange.start <= end)
                || (mapedFromRange.end >= start && mapedFromRange.end <= end)) {
                const _start = mapedToRange.start;
                const _end = mapedToRange.end;
                return {
                    data: data,
                    start: Math.min(_start, _end),
                    end: Math.max(_start, _end),
                };
            }
        }
    }
}
exports.SourceMapBase = SourceMapBase;
class SourceMap extends SourceMapBase {
    constructor(sourceDocument, mappedDocument, _mappings) {
        super(_mappings);
        this.sourceDocument = sourceDocument;
        this.mappedDocument = mappedDocument;
        this._mappings = _mappings;
    }
    getSourceRange(start, end, filter) {
        for (const maped of this.getRanges(start, end !== null && end !== void 0 ? end : start, false, filter)) {
            return maped;
        }
    }
    getMappedRange(start, end, filter) {
        for (const maped of this.getRanges(start, end !== null && end !== void 0 ? end : start, true, filter)) {
            return maped;
        }
    }
    getSourceRanges(start, end, filter) {
        return this.getRanges(start, end !== null && end !== void 0 ? end : start, false, filter);
    }
    getMappedRanges(start, end, filter) {
        return this.getRanges(start, end !== null && end !== void 0 ? end : start, true, filter);
    }
    *getRanges(start, end, sourceToTarget, filter) {
        const startIsNumber = typeof start === 'number';
        const endIsNumber = typeof end === 'number';
        const toDoc = sourceToTarget ? this.mappedDocument : this.sourceDocument;
        const fromDoc = sourceToTarget ? this.sourceDocument : this.mappedDocument;
        const startOffset = startIsNumber ? start : fromDoc.offsetAt(start);
        const endOffset = endIsNumber ? end : fromDoc.offsetAt(end);
        for (const maped of super.getRanges(startOffset, endOffset, sourceToTarget, filter)) {
            yield getMaped(maped);
        }
        function getMaped(maped) {
            if (startIsNumber) {
                return maped;
            }
            return {
                data: maped.data,
                start: toDoc.positionAt(maped.start),
                end: toDoc.positionAt(maped.end),
            };
        }
    }
}
exports.SourceMap = SourceMap;
